### PR TITLE
Raffle Submission

### LINKED ISSUE #5: Raffle Submission Page for Participants
We need to create a raffle submission page for participants, where they can see the questions, fill the form and submit.

### PR DESCRIPTION
**Pull request by ErtyumPX**
Closes #5 
Touches the first and the weakest protection way in #6 

## What are we doing in this PR?

- Creating `<Form/>`, `<Question/>` and `<AfterSubmission/>` components to be used in raffle submission page.
- Adjusting necessary communications between these components and the raffle submission page.
- Creates **placeholder** interfaces for `RaffleForm` and `Question` data which is passed to the related components, which will be adjusted after we aggree on the exact form of these objects.
- Assumes _kind of_ pseudo endpoints that we will use the get raffle information from. The parts that uses these functions with the `APIService` are commented out to be used after backend is ready.
- Creates a test form class to be used for checking the styles and behaviours in the fronted.

---

**Form Component:** Used directly in the `raffle[id]/page.tsx` and wraps the actual form component in it. Takes a raffle info object and a function to call necessary function in the `page.tsx`. Cannot be submitted if there is an invalid or empty field which is tagged as required. Gives an warning message when it happens.

**Question Component:** Used in the `Form` component, to create a new field for each question. Takes the index of the question in the form, question object and a callback function to notify the form when the value of the input field is changed. Tests the field value using specified regex if the field is tagged as required. If so and if the input is invalid or empty, shows a warning message.

**After Submission Component:** A simple component That is used directly in the `page.tsx` that is showed after submission. If the submission is recorded in local storage, form is not showed and directly this component appears.

## How does the UI look like?

Image of the beginning and the end of the test form:

screenshot-2024-03-27_21:01:55

---

screenshot-2024-03-27_21:02:05

## What will be done as follow-up?

`RaffleForm` and `Question` objects needs to be worked on to aggree on how will they used and what will be stored in them.

Content of the `AfterSubmission` component might also be passed in the `RaffleForm`.

### CODE REVIEW

**Comment by HuseyinSimsek7904** [comment]
I will review this file after we decide on the final structure of the raffle questions.

**Comment by ErtyumPX** [comment]
@HuseyinSimsek7904 the structure of the questions are partially irrelevant for this PR, as long as we know there will be title, description and types. We can change the question and how we will show the stuff later in a follow-up PR.

This PR is mostly introducing the way we are generation UI for each question there is in the array, the most of the `RaffleForm` structure will probably change as well

**Review by Utkub24 — APPROVED** [review]
>@HuseyinSimsek7904 the structure of the questions are partially irrelevant for this PR, as long as we know there will be title, description and types. We can change the question and how we will show the stuff later in a follow-up PR.

I agree with @ErtyumPX here, this PR could serve as a starter for the general feel and look of this page. Since the question forms are dependent on user input I believe we'll  procuderally generate this page in the future.  The input validation methods need a bit of work though!

File: src/interfaces/raffleForm.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  template: string;
+  isRequired: boolean;
+}
+
+export interface RaffleForm {
+  id: string;
+  title: string;
+  description: string;
+  questions: Question[];
+}
```
Comment (Utkub24) [inline src/interfaces/raffleForm.ts:14]: Should these interfaces be in separate files according to convention?

Comment (ErtyumPX) [reply src/interfaces/raffleForm.ts:14]: Yeah actually, will change that :+1:

Comment (ErtyumPX) [reply src/interfaces/raffleForm.ts]: Done

**Review by HuseyinSimsek7904 — APPROVED** [review]
Good work!

**Comment by ErtyumPX** [comment]
Are we good to merge this @HuseyinSimsek7904 :))

**Comment by HuseyinSimsek7904** [comment]
What am I supposed to test? Like which URL should I write to test this PR?

**Comment by HuseyinSimsek7904** [comment]
When I write some text that is not a number to the number field, it says "This field is required". Can we change that to something like  "This field requires a whole number"?

**Comment by ErtyumPX** [comment]
There is only one thing you can do, right after entering a random ID, you should be able to test the PR pal ^^

**Comment by ErtyumPX** [comment]
@HuseyinSimsek7904 more specific warning and error will be added in the `form-and-questions` branch :+1:

**Comment by HuseyinSimsek7904** [comment]
> @HuseyinSimsek7904 more specific warning and error will be added in the `form-and-questions` branch 👍

Okay then. Good work!

**Pull request merged**